embedchain/loaders/github.py

```diff
@@ -2,116 +2,258 @@
-    data = []
-    data_content = []
-    try:
-        with open(path, "rb") as f:
-            content = f.read().decode("utf-8")
-    except Exception as e:
-        print(f"Error reading file {path}: {e}")
-        raise ValueError(f"Failed to read file {path}")
-
-    meta_data = {}
-    meta_data["url"] = path
-    data.append(
-        {
-            "content": content,
-            "meta_data": meta_data,
-        }
-    )
-    data_content.append(content)
-    doc_id = hashlib.sha256((" ".join(data_content) + path).encode()).hexdigest()
-    return {
-        "doc_id": doc_id,
-        "data": data,
-    }
+from embedchain.utils import clean_string
+
+GITHUB_URL = "https://github.com"
+GITHUB_API_URL = "https://api.github.com"
+
+VALID_SEARCH_TYPES = set(["code", "repo", "pr", "issue"])
 
 
 class GithubLoader(BaseLoader):
-    def load_data(self, repo_url):
-        """Load data from a git repo."""
+    """Load data from github search query."""
+
+    def __init__(self, config: Optional[Dict[str, Any]] = None):
+        super().__init__()
+        if not config:
+            logging.warning("GithubLoader config is empty, client will not be initialized.")
+            return
+
         try:
-            from git import Repo
+            from github import Github
         except ImportError as e:
             raise ValueError(
-                "GithubLoader requires extra dependencies. Install with `pip install --upgrade 'embedchain[git]'`"
+                "GithubLoader requires extra dependencies. Install with `pip install --upgrade 'embedchain[github]'`"
             ) from e
 
-        mdx_loader = MdxLoader()
-        json_loader = JSONLoader()
+        self.config = config
+        token = config.get("token")
+        if not token:
+            raise ValueError(
+                "GithubLoader requires a personal access token to use github api. Check - `https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-personal-access-token-classic`"  # noqa: E501
+            )
+
+        try:
+            self.client = Github(token)
+        except Exception as e:
+            logging.error(f"GithubLoader failed to initialize client: {e}")
+            self.client = None
+
+    def _github_search_code(self, query: str):
+        """Search github code."""
+        data = []
+        results = self.client.search_code(query)
+        urls = []
+        for result in tqdm.tqdm(results, total=results.totalCount, desc="Loading code files from github"):
+            url = result.html_url
+            urls.append(url)
+            content = result.decoded_content.decode("utf-8")
+            metadata = {
+                "url": url,
+            }
+            data.append(
+                {
+                    "content": clean_string(content),
+                    "meta_data": metadata,
+                }
+            )
+        logging.info(f"Added data from urls: {urls}")
```

Comment:
The list of URLs can be really large, right? Might not be a good idea to print the list of urls altogether.